src/uiHelpers_nanos.c

```diff
@@ -0,0 +1,213 @@
+{
+	paginatedTextState_t* ctx = paginatedTextState;
+	assert_uiPaginatedText_magic();
+	ASSERT(ctx->currentText[SIZEOF(ctx->currentText) - 1] == '\0');
+	ASSERT(ctx->scrollIndex + SIZEOF(ctx->currentText) <= SIZEOF(ctx->fullText));
+	os_memmove(
+	        ctx->currentText,
+	        ctx->fullText + ctx->scrollIndex,
+	        SIZEOF(ctx->currentText) - 1
+	);
+	UX_REDISPLAY();
+}
+
+static void scroll_left()
+{
+	paginatedTextState_t* ctx = paginatedTextState;
+	assert_uiPaginatedText_magic();
+	if (ctx->scrollIndex > 0) {
+		ctx->scrollIndex--;
+		scroll_update_display_content();
+	}
+}
+
+static void scroll_right()
+{
+	paginatedTextState_t* ctx = paginatedTextState;
+	assert_uiPaginatedText_magic();
+	if (ctx->scrollIndex + SIZEOF(ctx->currentText) < 1 + strlen(ctx->fullText)) {
+		paginatedTextState->scrollIndex++;
+		scroll_update_display_content();
+	}
+}
+
+unsigned int ui_paginatedText_button(
+        unsigned int button_mask,
+        unsigned int button_mask_counter MARK_UNUSED
+)
+{
+	paginatedTextState_t* ctx = paginatedTextState;
+	TRY_CATCH_UI({
+		assert_uiPaginatedText_magic();
+		ASSERT(io_state == IO_EXPECT_UI);
+		ASSERT(device_is_unlocked() == true);
+		switch (button_mask)
+		{
+		case BUTTON_LEFT:
+		case BUTTON_EVT_FAST | BUTTON_LEFT: // SEEK LEFT
+			scroll_left();
+			break;
+
+		case BUTTON_RIGHT:
+		case BUTTON_EVT_FAST | BUTTON_RIGHT: // SEEK RIGHT
+			scroll_right();
+			break;
+
+		case BUTTON_EVT_RELEASED | BUTTON_LEFT | BUTTON_RIGHT: // PROCEED
+			uiCallback_confirm(&ctx->callback);
+			break;
+		}
+	});
+	return 0;
+}
+
+unsigned int ui_confirm_button(
+        unsigned int button_mask,
+        unsigned int button_mask_counter MARK_UNUSED
+)
+{
+	TRY_CATCH_UI({
+		assert_uiPrompt_magic();
+		ASSERT(io_state == IO_EXPECT_UI);
+		ASSERT(device_is_unlocked() == true);
+		switch (button_mask)
+		{
+		case BUTTON_EVT_RELEASED | BUTTON_LEFT: // REJECT
+			uiCallback_reject(&promptState->callback);
+			break;
+
+		case BUTTON_EVT_RELEASED | BUTTON_RIGHT: // APPROVE
+			uiCallback_confirm(&promptState->callback);
+			break;
+		}
+	})
+	return 0;
+}
+
+static const bagl_element_t ui_paginatedText[] = {
+	UI_BACKGROUND(),
+	UI_ICON_LEFT(ID_ICON_GO_LEFT, BAGL_GLYPH_ICON_LEFT),
+	UI_ICON_RIGHT(ID_ICON_GO_RIGHT, BAGL_GLYPH_ICON_RIGHT),
+
+	// TODO(ppershing): what are the following magical numbers?
+
+	UI_TEXT(ID_UNSPECIFIED, 0, 12, 128, &displayState.paginatedText.header),
+	UI_TEXT(ID_UNSPECIFIED, 0, 26, 128, &displayState.paginatedText.currentText),
+	#ifdef HEADLESS
+	HEADLESS_UI_ELEMENT(),
+	#endif
+};
+
+static const bagl_element_t* ui_prepro_paginatedText(const bagl_element_t *element)
+{
+	paginatedTextState_t* ctx = paginatedTextState;
+	assert_uiPaginatedText_magic();
+
+	bool full_text_overflows_screen = strlen(ctx->fullText) > strlen(ctx->currentText);
+	switch (element->component.userid) {
+	case ID_ICON_GO_LEFT:
+		return (ctx->scrollIndex == 0 && full_text_overflows_screen)
```

Comment:
Maybe I would reverset the logic to `(textFirstScreen || ctx->scrollIndex != 0) ? element : NULL` (change the order so that null is in the else clause